In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split

In [15]:
# Creating a dataset with 4 features
X,y = make_regression(n_samples = 100,n_informative = 4, n_features=4, noise = 20, random_state=13)

In [16]:
# Splitting the model
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size= 0.2, random_state=42)

In [17]:
# Testing with Linear Regression

from sklearn.linear_model import LinearRegression

lnr = LinearRegression()
lnr.fit(X_train,y_train)

lnr.predict(X_test)

print('Coefficients:', lnr.coef_)
print('Intercept : ', lnr.intercept_)

Coefficients: [94.51137102 34.57746238 62.63782903 80.89917745]
Intercept :  -2.3384147541578377


In [18]:
# Testing with ElasticNet Regression
from sklearn.linear_model import ElasticNet

en = ElasticNet(alpha =1, l1_ratio=0.5)

en.fit(X_train, y_train)
en.predict(X_test)

print("Coefficients:", en.coef_)
print("Intercept:", en.intercept_)

Coefficients: [57.97346777 16.92631463 29.96734784 49.62913425]
Intercept: -4.485717900282069


In [19]:
# Now, Trying with different l1 ratios

l1_ratios = [0, 0.25, 0.5, 0.75, 0.9, 1]
results = []

for ratio in l1_ratios:
    model = ElasticNet(alpha=1, l1_ratio=ratio)
    model.fit(X_train, y_train)

    results.append([ratio, model.coef_[0], model.coef_[1], model.coef_[2], model.coef_[3]])

df = pd.DataFrame(results, columns=['L1 Ratio','Feature 1','Feature 2', 'Feature 3', 'Feature 4'])

df.style.format("{:.4f}")

,L1 Ratio,Feature 1,Feature 2,Feature 3,Feature 4
0,0.0000,42.6151,11.1639,19.3711,36.5995
1,0.2500,49.0628,13.4681,23.5837,42.0729
2,0.5000,57.9735,16.9263,29.9673,49.6291
3,0.7500,71.1843,22.5814,40.5667,60.8247
4,0.9000,82.7856,28.0023,50.9102,70.6592
5,1.0000,93.1266,33.1422,60.8750,79.4328


**Observation:** 
> Keeping ``alpha = 1`` constant, increasing ``l1_ratio`` shifts ElasticNet from Ridge (L2) toward Lasso (L1). In this experiment, the coefficient magnitudes increased as the L1 contribution increased. This shows that changing the L1/L2 balance can significantly affect the learned coefficients.

> However, *this does not mean that Lasso always produces larger coefficients than Ridge. The result depends on the dataset and the chosen regularization strength.*

In [22]:
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import r2_score

models = {
    "Linear Regression": LinearRegression(),
    "Ridge": Ridge(alpha=1),
    "Lasso": Lasso(alpha=1),
    "ElasticNet": ElasticNet(alpha=1, l1_ratio=0.5)
}

results = []

for name, model in models.items():

    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)

    results.append([
        name,
        r2_score(y_test, y_pred)
    ])

df = pd.DataFrame(
    results,
    columns=["Model", "R2_Score"]
)

df.style.format({"R2_Score": "{:.4f}"})

,Model,R2_Score
0,Linear Regression,0.9781
1,Ridge,0.9765
2,Lasso,0.9760
3,ElasticNet,0.7646


**Observation:**

> In this experiment, Linear Regression achieved the highest **R2 score (0.9781)**, followed closely by **Ridge (0.9765)** and **Lasso (0.9760)**. ElasticNet with ``alpha=1`` and ``l1_ratio=0.5`` performed significantly worse, with an R2 of **0.7646**. This does not mean ElasticNet is inherently worse. 

> It indicates that the chosen ElasticNet hyperparameters were not suitable for this dataset. Therefore, alpha and l1_ratio should be tuned to find a suitable combination.